File: charts/lagoon-core/files/seed-core.py
import os
import requests

TOKEN = os.getenv('LEGACY_TOKEN')
INITIAL_USER_EMAIL = os.getenv('INITIAL_USER_EMAIL')
INITIAL_USER_PASSWORD = os.getenv('INITIAL_USER_PASSWORD')
INITIAL_ORGANIZATION_NAME = os.getenv('INITIAL_ORGANIZATION_NAME')

KEYCLOAK_ADMIN_API_CLIENT_SECRET = os.getenv('KEYCLOAK_ADMIN_API_CLIENT_SECRET')
KEYCLOAK_FRONTEND_URL = os.getenv('KEYCLOAK_FRONTEND_URL')

LAGOON_API_URL = os.getenv('LAGOON_API_URL')

seed_user_gql = f"""
mutation LagoonCoreSeeding {{

    CreatePlatformOwner: addUser(
        input: {{
            email: "{INITIAL_USER_EMAIL}"
        }}
    ) {{
        id
    }}

    GiveUserPlatformOwnerRole: addPlatformRoleToUser(
    user: {{
      email: "{INITIAL_USER_EMAIL}"
    }}
    role: OWNER
    ) {{
        email
        platformRoles
    }}
}}
"""

seed_org_gql = f"""
mutation LagoonCoreSeedOrg {{
    CreateOrganization: addOrganization(input: {{
        name: "{INITIAL_ORGANIZATION_NAME}"
    }}) {{
        id
    }}
}}
"""


def seed_org():
    print("Seeding Lagoon core with initial organization")

    lagoon_api_url = LAGOON_API_URL.replace("https://", "http://", 1)
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {TOKEN}"
    }

    response = requests.post(lagoon_api_url, json={"query": seed_org_gql}, headers=headers)
    body = response.json()
    print(body)
    try:
        org_id = body["data"]["CreateOrganization"]["id"]
        print(f"Created org with id {org_id}")
    except (KeyError, TypeError):
        print("Failed to extract organization ID from response:")
        print(body)
        return

    configure_org_graphql = f"""
mutation LagoonCoreConfigOrg {{
    CreateGroup: addGroup(input: {{
        name: "{INITIAL_ORGANIZATION_NAME}-group"
    }}) {{
        id
    }}

    AddUserToGroup: addUserToGroup(input: {{
        user: {{ email: "{INITIAL_USER_EMAIL}" }}
        group: {{ name: "{INITIAL_ORGANIZATION_NAME}-group" }}
        role: OWNER
    }}) {{
        name
    }}

    AddGroupToOrg: addExistingGroupToOrganization(input: {{
        name: "{INITIAL_ORGANIZATION_NAME}-group"
        organization: {org_id}
    }}) {{
        id
    }}

    AddUserToOrg: addAdminToOrganization(input: {{
        user: {{ email: "{INITIAL_USER_EMAIL}" }}
        organization: {{ id: {org_id} }}
        role: OWNER
    }}) {{
        id
    }}
}}
"""
    configure_response = requests.post(lagoon_api_url, json={"query": configure_org_graphql}, headers=headers)
    print(configure_response.json())


def seed_user():
    print("Seeding Lagoon core with initial user")

    lagoon_api_url = LAGOON_API_URL.replace("https://", "http://", 1)
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {TOKEN}"
    }

    response = requests.post(lagoon_api_url, json={"query": seed_user_gql}, headers=headers)
    print(response.json())

    keycloak_url = f"{KEYCLOAK_FRONTEND_URL}/auth".replace("https://", "http://", 1)
    print(f"Setting user password in keycloak at {keycloak_url}")

    data = {
            "grant_type": "client_credentials",
            "client_id": "admin-api",
            "client_secret": KEYCLOAK_ADMIN_API_CLIENT_SECRET
            }
    response = requests.post(f"{keycloak_url}/realms/master/protocol/openid-connect/token", data)
    if response.status_code != 200:
        print("Failed to get Keycloak access token")
        return
    else:
        print("Keycloak token successfully retrieved")

    token = response.json()["access_token"]
    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json"
    }

    user_lookup = requests.get(f"{keycloak_url}/admin/realms/lagoon/users?email={INITIAL_USER_EMAIL}", headers=headers)
    if user_lookup.status_code != 200:
        print(f"Failed to user {INITIAL_USER_EMAIL} in Keycloak")
        return

    users = user_lookup.json()
    user_id = users[0]["id"]
    pw_data = {
        "type": "password",
        "value": INITIAL_USER_PASSWORD,
        "temporary": False
    }

    reset_pw = requests.put(
        f"{keycloak_url}/admin/realms/lagoon/users/{user_id}/reset-password",
        headers=headers,
        json=pw_data
    )

    if reset_pw.status_code == 204:
        print("Password set successfully!")
    else:
        print("Failed to set user password.")


seed_user()
if INITIAL_ORGANIZATION_NAME != "":
    seed_org()

File: charts/lagoon-core/files/seed-core.js
const axios = require("axios");
const jwt = require("jsonwebtoken");

const {
  INITIAL_USER_EMAIL,
  INITIAL_USER_PASSWORD,
  INITIAL_ORGANIZATION_NAME,
  KEYCLOAK_ADMIN_API_CLIENT_SECRET,
  KEYCLOAK_FRONTEND_URL,
  LAGOON_API_URL,
  JWTUSER,
  JWTAUDIENCE,
  JWTSECRET
} = process.env;

const payload = {
  role: "admin",
  iss: JWTUSER,
  aud: JWTAUDIENCE,
  sub: JWTUSER
};

const TOKEN = jwt.sign(payload, JWTSECRET, { algorithm: "HS256" });

const seedUserGql = `
mutation LagoonCoreSeeding {
    CreatePlatformOwner: addUser(
        input: { email: "${INITIAL_USER_EMAIL}" }
    ) {
        id
    }

    GiveUserPlatformOwnerRole: addPlatformRoleToUser(
        user: { email: "${INITIAL_USER_EMAIL}" }
        role: OWNER
    ) {
        email
        platformRoles
    }
}
`;

const seedOrgGql = `
mutation LagoonCoreSeedOrg {
    CreateOrganization: addOrganization(input: {
        name: "${INITIAL_ORGANIZATION_NAME.toLocaleLowerCase().replace(/[^0-9a-z-]/g,'-')}"
        friendlyName: "${INITIAL_ORGANIZATION_NAME}"
    }) {
        id
    }
}
`;

async function waitForKeycloak(url, timeout = 600000, interval = 2000) {
  const start = Date.now();

  while (Date.now() - start < timeout) {
    try {
      const response = await axios.get(url);
      if (response.status === 200) {
        console.log("Connected to keycloak");
        return;
      }
    } catch (err) {
      console.log(`waiting for keycloak at ${url}...`);
    }
    await new Promise(resolve => setTimeout(resolve, interval));
  }

  throw new Error(`Timed out waiting for Keycloak at ${url}`);
}

async function seedOrg() {
  console.log("Seeding Lagoon core with initial organization");

  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${TOKEN}`
  };

  try {
    const response = await axios.post(LAGOON_API_URL, { query: seedOrgGql }, { headers });
    const body = response.data;

    const orgId = body?.data?.CreateOrganization?.id;
    if (!orgId) {
      console.log("Failed to extract organization ID from response:");
      console.log(body);
      return;
    }

    console.log(`Created org with id ${orgId}`);
  } catch (err) {
    console.error("Error seeding organization:", err.response?.data || err.message);
  }
}

async function seedUser() {
  console.log("Seeding Lagoon core with initial user");

  const headers = {
    "Content-Type": "application/json",
    Authorization: `Bearer ${TOKEN}`
  };

  try {
    const response = await axios.post(LAGOON_API_URL, { query: seedUserGql }, { headers });

    await waitForKeycloak(KEYCLOAK_FRONTEND_URL + "/admin/master/console");
    console.log(`Setting user password in keycloak at ${KEYCLOAK_FRONTEND_URL}`);

    const data = new URLSearchParams({
      grant_type: "client_credentials",
      client_id: "admin-api",
      client_secret: KEYCLOAK_ADMIN_API_CLIENT_SECRET
    });

    const tokenResponse = await axios.post(
      `${KEYCLOAK_FRONTEND_URL}/realms/master/protocol/openid-connect/token`,
      data
    );

    if (tokenResponse.status !== 200) {
      console.log("Failed to get Keycloak access token");
      return;
    } else {
      console.log("Keycloak token successfully retrieved");
    }

    const token = tokenResponse.data.access_token;
    const keycloakHeaders = {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json"
    };

    const userLookup = await axios.get(
      `${KEYCLOAK_FRONTEND_URL}/admin/realms/lagoon/users?email=${INITIAL_USER_EMAIL}`,
      { headers: keycloakHeaders }
    );

    if (userLookup.status !== 200) {
      console.log(`Failed to find user ${INITIAL_USER_EMAIL} in Keycloak`);
      return;
    }

    const users = userLookup.data;
    const userId = users[0]?.id;
    if (!userId) {
      console.log("User not found in Keycloak");
      return;
    }

    const pwData = {
      type: "password",
      value: INITIAL_USER_PASSWORD,
      temporary: false
    };

    const resetPw = await axios.put(
      `${KEYCLOAK_FRONTEND_URL}/admin/realms/lagoon/users/${userId}/reset-password`,
      pwData,
      { headers: keycloakHeaders }
    );

    if (resetPw.status === 204) {
      console.log("Password set successfully!");
    } else {
      console.log("Failed to set user password.");
    }
  } catch (err) {
    console.error("Error seeding user:", err.response?.data || err.message);
  }
}

(async () => {
  if (INITIAL_USER_EMAIL && INITIAL_USER_PASSWORD) {
    await seedUser();
  }
  if (INITIAL_ORGANIZATION_NAME) {
    await seedOrg();
  }
})();

